compiler/rustc_lint/src/types.rs

```diff
@@ -1241,15 +1446,80 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
+                }]);
+            }
+            ffir @ FfiResult::FfiUnsafeWrapper { .. } => {
+                let mut last_ty = None;
+                let mut ffiresult_recursor = Some(&ffir);
+                let mut cimproper_layers: Vec<ImproperCTypesLayer<'tcx>> = vec![];
+
+                // this whole while block converts the arbitrarily-deep
+                // FfiResult stack to a ImproperCTypesLayer Vec
+                while let Some(ref ffir_rec) = ffiresult_recursor {
+                    match ffir_rec {
+                        FfiResult::FfiPhantom(ty) => {
+                            last_ty = Some(ty.clone());
+                            let len = cimproper_layers.len();
+                            if len > 0 {
+                                cimproper_layers[len - 1].inner_ty = last_ty.clone();
+                            }
+                            cimproper_layers.push(ImproperCTypesLayer {
+                                ty: ty.clone(),
+                                inner_ty: None,
+                                help: None,
+                                note: fluent::lint_improper_ctypes_only_phantomdata,
+                                span_note: None, // filled later
+                            });
+                            ffiresult_recursor = None;
+                        }
+                        FfiResult::FfiUnsafe { ty, reason, help }
+                        | FfiResult::FfiUnsafeWrapper { ty, reason, help, .. } => {
+                            last_ty = Some(ty.clone());
+                            let len = cimproper_layers.len();
+                            if len > 0 {
+                                cimproper_layers[len - 1].inner_ty = last_ty.clone();
+                            }
+                            cimproper_layers.push(ImproperCTypesLayer {
+                                ty: ty.clone(),
+                                inner_ty: None,
+                                help: help.clone(),
+                                note: reason.clone(),
+                                span_note: None, // filled later
+                            });
+
+                            if let FfiResult::FfiUnsafeWrapper { wrapped, .. } = ffir_rec {
+                                ffiresult_recursor = Some(wrapped.as_ref());
+                            } else {
+                                ffiresult_recursor = None;
+                            }
+                        }
+                        FfiResult::FfiSafe => {
+                            bug!("malformed FfiResult stack: it should be unsafe all the way down")
+                        }
+                    };
+                }
+                let last_ty = match last_ty {
+                    Some(last_ty) => last_ty,
+                    None => bug!(
+                        "This option should definitely have been filled by the loop that just finished"
+                    ),
+                };
```

Comment:
I believe we can skip always cloning the `last_ty` this way. Now it's, er, just the last `ty`.
```suggestion
                // should always have at least one type
                let last_ty = cimproper_layers.last().unwrap().ty.clone();
```